<?xml version="1.0" encoding="utf-8"?>
<launch>
  <arg name="odom_topic" value="/visual_slam/odom" />
  <arg name="depth_topic" value="/pcl_render_node/depth" />
  <arg name="global_env_pcd2_topic" value="/random_forest/all_map" />
  <arg name="map_size_x" value="14.0" />
  <arg name="map_size_y" value="70.0" />
  <arg name="map_size_z" value="3.5" />
  <arg name="origin_x" value=" -7.0" />
  <arg name="origin_y" value=" -35.0" />
  <arg name="origin_z" value=" -1.0" />
  <arg name="resolution" value="0.1" />
  <arg name="vel_limit" value="7.0" />
  <arg name="acc_limit" value="5.0" />
  <arg name="jerk_limit" value="15.0" />
  <arg name="rho_time" value="0.01" />
  <arg name="global_test" value="true" />

  <node pkg="state_machine" type="test_planner_node" name="state_machine_node" output="screen">

    <remap from="/global_cloud" to="$(arg global_env_pcd2_topic)"/>
    <remap from="/odom_topic" to="$(arg odom_topic)"/>
    <remap from="/depth_topic" to="$(arg depth_topic)" />

    <param name="occ_map/origin_x" value="$(arg origin_x)" type="double"/>
    <param name="occ_map/origin_y" value="$(arg origin_y)" type="double"/>
    <param name="occ_map/origin_z" value="$(arg origin_z)" type="double"/>
    <param name="occ_map/map_size_x" value="$(arg map_size_x)" type="double"/>
    <param name="occ_map/map_size_y" value="$(arg map_size_y)" type="double"/>
    <param name="occ_map/map_size_z" value="$(arg map_size_z)" type="double"/>
    <param name="occ_map/local_radius_x" value="20" />
    <param name="occ_map/local_radius_y" value="40" />
    <param name="occ_map/local_radius_z" value="4" />
    <param name="occ_map/resolution" value="$(arg resolution)" type="double"/>
    <param name="occ_map/show_raw_depth" value="true" type="bool"/>
    <param name="occ_map/show_filter_proj_depth" value="true" type="bool"/>
    <param name="occ_map/depth_scale" value="1000.0"/>
    <param name="occ_map/depth_filter_maxdist" value="6.0"/>
    <param name="occ_map/depth_filter_mindist" value="0.1"/>
    <param name="occ_map/depth_filter_margin" value="1"/>
    <param name="occ_map/skip_pixel" value="2"/>

    <!-- use when mapping frequency is high (>= 30Hz) -->
    <param name="occ_map/use_shift_filter" value="false" type="bool"/>
    <param name="occ_map/depth_filter_tolerance" value="0.2"/>

    <param name="occ_map/cx" value="257.296"/>
    <param name="occ_map/cy" value="236.743"/>
    <param name="occ_map/fx" value="385.754"/>
    <param name="occ_map/fy" value="385.754"/>
    <param name="occ_map/rows" value="480"/>
    <param name="occ_map/cols" value="640"/>

    <param name="occ_map/prob_hit_log" value="1.2"/>
    <param name="occ_map/prob_miss_log" value="-0.30"/>
    <param name="occ_map/clamp_min_log" value="-1.0"/>
    <param name="occ_map/clamp_max_log" value="2.0"/>
    <param name="occ_map/min_occupancy_log" value="1.39"/>
    <param name="occ_map/min_ray_length" value="0.1"/>
    <param name="occ_map/max_ray_length" value="6.0"/>
    <param name="occ_map/use_global_map" value="$(arg global_test)" type="bool"/>
    <param name="occ_map/inflate_length" value="0.19"/>

    <param name="pos_checker/hrz_safe_radius" value="0.4"/>
    <param name="pos_checker/vtc_safe_radius" value="0.2"/>
    <param name="pos_checker/copter_diag_len" value="0.4"/>
    <param name="pos_checker/dt" value="0.02"/>
    <param name="pos_checker/inflate" value="true"/>
    <param name="pos_checker/global_test" value="$(arg global_test)"/>

    <!-- kinorrt* params -->
    <param name="krrt/rho" value="$(arg rho_time)" type="double"/>
    <!-- the quadratic matrix R of u'Ru -->
    <param name="krrt/vel_limit" value="$(arg vel_limit)" type="double" />
    <param name="krrt/acc_limit" value="$(arg acc_limit)" type="double" />
    <param name="krrt/jerk_limit" value="$(arg jerk_limit)" type="double" />
    <param name="krrt/tree_node_nums" value="5000" type="int" />
    <param name="krrt/orphan_nums" value="1500" type="int" />
    <param name="krrt/debug_vis" value="false" type="bool" />
    <param name="krrt/radius_cost_between_two_states" value="4" type="double" />
    <!-- cost tolerance -->
    <param name="krrt/allow_orphan" value="false" type="bool" />
    <param name="krrt/allow_close_goal" value="false" type="bool" />
    <param name="krrt/stop_after_first_traj_found" value="false" type="bool" />
    <param name="krrt/rewire" value="true" type="bool" />
    <param name="krrt/use_regional_opt" value="false" type="bool" />
    <param name="krrt/use_deform" value="false" type="bool" />
    <param name="krrt/test_convergency" value="true" type="bool" />
    <param name="krrt/min_deform_duration" value="0.8" type="double" />

    <param name="bikrrt/rho" value="$(arg rho_time)" type="double"/>
    <!-- the quadratic matrix R of u'Ru -->
    <param name="bikrrt/vel_limit" value="$(arg vel_limit)" type="double" />
    <param name="bikrrt/acc_limit" value="$(arg acc_limit)" type="double" />
    <param name="bikrrt/jerk_limit" value="$(arg jerk_limit)" type="double" />
    <param name="bikrrt/tree_node_nums" value="5000" type="int" />
    <param name="bikrrt/orphan_nums" value="1500" type="int" />
    <param name="bikrrt/debug_vis" value="false" type="bool" />
    <param name="bikrrt/radius_cost_between_two_states" value="4" type="double" />
    <!-- cost tolerance -->
    <param name="bikrrt/allow_orphan" value="false" type="bool" />
    <param name="bikrrt/allow_close_goal" value="false" type="bool" />
    <param name="bikrrt/stop_after_first_traj_found" value="false" type="bool" />
    <param name="bikrrt/rewire" value="true" type="bool" />
    <param name="bikrrt/use_regional_opt" value="false" type="bool" />
    <param name="bikrrt/use_deform" value="false" type="bool" />
    <param name="bikrrt/test_convergency" value="true" type="bool" />

    <param name="search/max_tau" value="0.5" type="double"/>
    <param name="search/init_max_tau" value="0.8" type="double"/>
    <param name="search/max_vel" value="$(arg vel_limit)" type="double"/>
    <param name="search/max_acc" value="$(arg acc_limit)" type="double"/>
    <param name="search/w_time" value="100.0" type="double"/>
    <param name="search/horizon" value="600.0" type="double"/>
    <param name="search/lambda_heu" value="1" type="double"/>
    <param name="search/resolution_astar" value="$(arg resolution)" type="double"/>
    <param name="search/time_resolution" value="0.8" type="double"/>
    <param name="search/allocate_num" value="1000000" type="int"/>


    <param name="search/jerk/max_vel" value="$(arg vel_limit)" type="double"/>
    <param name="search/jerk/max_acc" value="$(arg acc_limit)" type="double"/>
    <param name="search/jerk/w_time" value="100.0" type="double"/>
    <param name="search/jerk/horizon" value="5.0" type="double"/>
    <param name="search/jerk/lambda_heu" value="3" type="double"/>
    <param name="search/jerk/resolution_astar" value="$(arg resolution)" type="double"/>
    <param name="search/jerk/allocate_num" value="1000000" type="int"/>
    <param name="search/jerk/max_jerk" value="2.0" type="double"/>
    <param name="search/jerk/rho" value="$(arg rho_time)" type="double"/>
    <param name="search/jerk/input_res" value="1.0" type="double"/>
    <param name="search/jerk/time_res" value="0.2" type="double"/>
    <param name="search/jerk/max_tau" value="2.5" type="double"/>
    <param name="search/jerk/input_res_int" value="0.125" type="double"/>
    <param name="search/jerk/time_res_init" value="0.125" type="double"/>
    <param name="search/jerk/init_max_tau" value="2.0" type="double"/>


    <param name="topo_prm/sample_inflate_x" value="1.0" type="double"/>
    <param name="topo_prm/sample_inflate_y" value="3.0" type="double"/>
    <param name="topo_prm/sample_inflate_z" value="1.0" type="double"/>
    <param name="topo_prm/clearance" value="0.3" type="double"/>
    <param name="topo_prm/max_sample_time" value="1.0" type="double"/>
    <param name="topo_prm/max_sample_num" value="2000" type="int"/>
    <param name="topo_prm/max_raw_path" value="300" type="int"/>
    <param name="topo_prm/max_raw_path2" value="300" type="int"/>
    <param name="topo_prm/reserve_num" value="1" type="int"/>
    <param name="topo_prm/ratio_to_short" value="5.5" type="double"/>
    <param name="topo_prm/parallel_shortcut" value="true" type="bool"/>

    <param name="r3/sampling_space_inflate" value="5.0" type="double"/>
    <param name="r3/ground_height" value="0.0" type="double"/>
    <param name="r3/ceiling_height" value="2.0" type="double"/>
    <param name="r3/search_time_per_meter" value="0.002" type="double"/>
    <param name="r3/lattice_step" value="0.3" type="double"/>
    <param name="r3/use_ompl" value="true" type="bool"/>

    <param name="rrt/steer_length" value="3.0" type="double"/>
    <param name="rrt/search_radius" value="5.0" type="double"/>
    <param name="rrt/search_time" value="1.0" type="double"/>
    <param name="rrt/max_tree_node_nums" value="5000" type="int"/>

    <param name="sampler/vel_mag_mean" value="3.5" type="double" />
    <param name="sampler/pos_hrz_var" value="5.5" type="double" />
    <param name="sampler/pos_vtc_var" value="5.0" type="double" />
    <param name="sampler/vel_mag_var" value="1.5" type="double" />
    <param name="sampler/vel_dir_var" value="0.3" type="double" />
    <param name="sampler/resolution" value="$(arg resolution)" type="double" />

    <param name="optimization/vel_limit" value="$(arg vel_limit)" type="double" />
    <param name="optimization/acc_limit" value="$(arg acc_limit)" type="double" />
    <param name="optimization/jerk_limit" value="$(arg jerk_limit)" type="double" />
    <param name="optimization/max_iter_times" value="30" type="int" />

    <param name="trunk_opt/rho" value="$(arg rho_time)" type="double" />
    <param name="trunk_opt/vel_limit" value="$(arg vel_limit)" type="double" />
    <param name="trunk_opt/acc_limit" value="$(arg acc_limit)" type="double" />
    <param name="trunk_opt/jerk_limit" value="$(arg jerk_limit)" type="double" />
    <param name="trunk_opt/discretization" value="10" type="int" />
    <param name="trunk_opt/rho_pos" value="100.0" type="double" />
    <param name="trunk_opt/rho_vel" value="100.0" type="double" />
    <param name="trunk_opt/rho_acc" value="100.0" type="double" />
    <param name="trunk_opt/dist_tolerance" value="0.1" type="double" />

    <param name="test_plan/start_x" value="5.0" type="double"/>
    <param name="test_plan/start_y" value="30.0" type="double"/>
    <param name="test_plan/start_z" value="0.0" type="double"/>
    <param name="test_plan/use_optimization" value="false" type="bool"/>
    <param name="test_plan/use_r3" value="true" type="bool"/>
    <param name="test_plan/replan_time" value="1.0" type="double"/>
    <param name="test_plan/test_variants" value="false" type="bool"/>
    <param name="test_plan/test_wo" value="true" type="bool"/>
    <param name="test_plan/test_hierarchical" value="false" type="bool"/>
    <param name="test_plan/use_bench" value="true" type="bool"/>
    <param name="test_plan/test_bikrrt_w" value="true" type="bool"/>
    <param name="test_plan/test_bikrrt_wo" value="true" type="bool"/>
    <param name="test_plan/test_rrt_w" value="true" type="bool"/>
    <param name="test_plan/test_rrt_wo" value="true" type="bool"/>
    <param name="test_plan/test_a_triple" value="true" type="bool"/>
    <param name="test_plan/repeat_times" value="1" type="int"/>
    <param name="test_plan/ctrl_pt_dist" value="0.5" type="double"/>
    <param name="test_plan/max_vel" value="$(arg vel_limit)" type="double"/>
    <param name="test_plan/max_acc" value="$(arg acc_limit)" type="double"/>
    <param name="test_plan/max_jerk" value="$(arg jerk_limit)" type="double"/>
    <param name="test_plan/file_name" value="/home/yehongkai/ros_ws/stdt_github_ws/src/std-trees/benchmark_data/wo-test1/" type="string"/>
    <param name="test_plan/rho" value="$(arg rho_time)" type="double"/>

    <param name="test_plan/print_debug_info" value="false" type="bool"/>
    <param name="test_plan/collision_check_dt" value="0.02" type="double"/>
    <param name="test_plan/max_loop_num" value="50" type="int"/>
    <param name="test_plan/bbox_width" value="7.0" type="double"/>

    <param name="test_plan/start_pos_x" value="0.0" type="double"/>
    <param name="test_plan/start_pos_y" value="34.0" type="double"/>
    <param name="test_plan/start_pos_z" value="0.0" type="double"/>
    <param name="test_plan/end_pos_x" value="0.0" type="double"/>
    <param name="test_plan/end_pos_y" value="-34.0" type="double"/>
    <param name="test_plan/end_pos_z" value="0.0" type="double"/>

  </node>

</launch>